sdk/search/Azure.Search.Documents/src/Batching/SearchIndexingBufferedSender{T}.cs

```diff
@@ -0,0 +1,606 @@
+        /// service.
+        /// </summary>
+        internal virtual SearchClient SearchClient { get; }
+
+        /// <summary>
+        /// Gets the URI endpoint of the Search Service.  This is likely
+        /// to be similar to "https://{search_service}.search.windows.net".
+        /// </summary>
+        /// <remarks>
+        /// This is not the URI of the Search Index.  You could construct that
+        /// URI with "{Endpoint}/indexes/{IndexName}" if needed.
+        /// </remarks>
+        public virtual Uri Endpoint => SearchClient.Endpoint;
+
+        /// <summary>
+        /// Gets the name of the Search Service.
+        /// </summary>
+        public virtual string ServiceName => SearchClient.ServiceName;
+
+        /// <summary>
+        /// Gets the name of the Search Index.
+        /// </summary>
+        public virtual string IndexName => SearchClient.IndexName;
+
+        /// <summary>
+        /// Gets a function that can be used to access the index key value of a
+        /// document.
+        /// </summary>
+        internal virtual Func<T, string> KeyFieldAccessor { get; private set; }
+
+        /// <summary>
+        /// Ensure we only lookup the key field accessor once, but that
+        /// everyone waits until we have one.
+        /// </summary>
+        private InitializationBarrierSlim _keyFieldBarrier = new InitializationBarrierSlim();
+
+        /// <summary>
+        /// Async event raised whenever an indexing action is added to the
+        /// sender.
+        /// </summary>
+        public event Func<IndexDocumentsAction<T>, CancellationToken, Task> ActionAddedAsync;
+
+        /// <summary>
+        /// Async event raised whenever an indexing action is sent by the
+        /// sender.
+        /// </summary>
+        public event Func<IndexDocumentsAction<T>, CancellationToken, Task> ActionSentAsync;
+
+        /// <summary>
+        /// Async event raised whenever an indexing action was submitted
+        /// successfully.
+        /// </summary>
+        public event Func<IndexDocumentsAction<T>, IndexingResult, CancellationToken, Task> ActionCompletedAsync;
+
+        /// <summary>
+        /// Async event raised whenever an indexing action failed.  The
+        /// <see cref="IndexingResult"/> or <see cref="Exception"/> may be null
+        /// depending on the failure.
+        /// </summary>
+        public event Func<IndexDocumentsAction<T>, IndexingResult, Exception, CancellationToken, Task> ActionFailedAsync;
+
+        /// <summary>
+        /// Creates a new instance of the SearchIndexingBufferedSender.
+        /// </summary>
+        /// <param name="searchClient">
+        /// The SearchClient used to send requests to the service.
+        /// </param>
+        /// <param name="options">
+        /// Provides the configuration options for the sender.
+        /// </param>
+        protected internal SearchIndexingBufferedSender(
```

Comment:
Mentioned above, but this should just have a protected ctor for mocking, and this could simply be internal.